### PR TITLE
FIX enable_adapters()/disable_adapter() unfreezing adapters loaded with inference_mode=True

### LINKED ISSUE #3447: disable_adapter() / enable_adapter_layers() silently unfreezes adapters loaded with inference_mode=True
### System Info

Reproduced against the current `main` branch source (`src/peft/tuners/tuners_utils.py`, `BaseTunerLayer.enable_adapters`), with `torch` installed locally, `get_peft_model` + `LoraConfig`.

### Who can help?

@benjaminbossan @githubnemo

### Reproduction

```python
import torch
import torch.nn as nn
from peft import get_peft_model, LoraConfig


class SmallModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin = nn.Linear(8, 8)

    def forward(self, x):
        return self.lin(x)


model = get_peft_model(
    SmallModel(),
    LoraConfig(target_modules=["lin"], r=4, lora_alpha=8, inference_mode=True),
)

def lora_requires_grad():
    return [p.requires_grad for n, p in model.named_parameters() if "lora_" in n]

print("Before:", lora_requires_grad())

with model.disable_adapter():
    print("Inside:", lora_requires_grad())

print("After:", lora_requires_grad())
```

Actual output:
```
Before: [False, False]
Inside: [False, False]
After: [True, True]
```

### Expected behavior

`requires_grad` should stay `False` after the `with model.disable_adapter():` block exits, matching the `inference_mode=True` the adapter was loaded with. Instead it flips to `True`, silently unfreezing an adapter that was explicitly configured as frozen.

### Root cause

`BaseTunerLayer.enable_adapters` (`src/peft/tuners/tuners_utils.py`, lines 1564-1574):

```python
def enable_adapters(self, enabled: bool) -> None:
    """Toggle the enabling and disabling of adapters
    Takes care of setting the requires_grad flag for the adapter weights.
    """
    if enabled:
        self.set_adapter(self.active_adapters)
        self._disable_adapters = False
    else:
        ...
```

`set_adapter` is called with no `inference_mode` argument, so it defaults to `False`, forcing `requires_grad=True` on every active adapter parameter regardless of the config the adapter was actually loaded/set with.

`PeftModel.disable_adapter()` (`src/peft/peft_model.py`, ~lines 1068-1084) calls `self.base_model.enable_adapter_layers()` in its `finally` block to restore adapters after temporarily disabling them, the standard pattern for computing a reference/base-model output (e.g. in DPO/PPO/RLHF-style training scripts). Every use of this context manager unconditionally re-enables `requires_grad=True` on the active adapter, even when it was intentionally frozen.

### Not a duplicate of #2928 / PR #2936

That issue/PR is about `modules_to_save`/`AuxiliaryTrainingWrapper` specifically (the *original*, non-adapter module's gradient state getting toggled), and its own fix only touches `_set_adapter` in `utils/other.py`. It doesn't touch `BaseTunerLayer.enable_adapters` in `tuners_utils.py`, the path exercised here, and the existing test added for that PR (`test_requires_grad_modules_to_save_disabling`) never constructs an adapter with `inference_mode=True`, so this path isn't covered.

### Impact

Any code that calls `model.disable_adapter()` (or `enable_adapter_layers()` directly) on a model containing an adapter that was explicitly frozen (`inference_mode=True`, e.g. a loaded reference adapter for RLHF-style training, or any adapter loaded with `is_trainable=False`) will have that adapter silently become trainable again as soon as the context manager exits. This can corrupt a "frozen reference" adapter if an optimizer or trainable-parameter filter is built after such a call, and makes `requires_grad` an unreliable signal for introspecting which parameters are actually meant to be trainable.

### PR DESCRIPTION
**Pull request by DaoyuanLi2816**
## What does this PR do?

Fixes #3447.

`BaseTunerLayer.enable_adapters` — invoked by `PeftModel.disable_adapter()`'s context-manager exit and by the public `enable_adapter_layers()`/`disable_adapter_layers()` methods — calls:

```python
self.set_adapter(self.active_adapters)
```

with no `inference_mode` argument, so it defaults to `False`. This unconditionally sets `requires_grad=True` on every active adapter, silently unfreezing an adapter that was explicitly loaded or set with `inference_mode=True` (e.g. a frozen reference adapter for RLHF-style training, or any adapter loaded with `is_trainable=False`).

### Reproduction (from the issue)

```python
model = get_peft_model(
    SmallModel(),
    LoraConfig(target_modules=["lin"], r=4, lora_alpha=8, inference_mode=True),
)
print([p.requires_grad for n, p in model.named_parameters() if "lora_" in n])  # [False, False]
with model.disable_adapter():
    pass
print([p.requires_grad for n, p in model.named_parameters() if "lora_" in n])  # [True, True] -- bug
```

While tracing the root cause (see [comment](https://github.com/huggingface/peft/issues/3447#issuecomment-5010032441)), I found the identical unconditional-unfreeze pattern independently duplicated in two more places, so this bundles fixes for all of them per this repo's `.ai/AGENTS.md` ("If an issue is small and affects multiple PEFT methods... fix all of them in the same PR"):

- `ModulesToSaveWrapper.enable_adapters` (`utils/other.py`) — same bug for `modules_to_save`.
- `LNTuningLayer.enable_adapters` (`tuners/ln_tuning/layer.py`) — a separate, full reimplementation with the same bug for LN Tuning.
- `TrainableTokensWrapper` and `UniLoraLayer` delegate their `enable_adapters` to a `BaseTunerLayer`-derived object, so they're fixed by the base-class change, but their signatures needed updating to accept and forward the new parameter.

Not a duplicate of #2928/PR #2936 (fixes `_set_adapter`/`ModulesToSaveWrapper.set_adapter`, a different function, and its test never constructs an adapter with `inference_mode=True`) or of PR #3290 (fixes `BaseTuner.inject_adapter`'s housekeeping call, a different, model-level method).

### Fix

`BaseTuner._enable_adapter_layers` now derives each module's active adapter(s)' own `inference_mode` from `peft_config` when re-enabling, instead of unconditionally marking them trainable, and passes it through `enable_adapters` -> `set_adapter`, which already handles `inference_mode` correctly (this mirrors the same `peft_config[name].inference_mode` lookup already used by `inject_adapter`'s housekeeping call). The two duplicated implementations (`ModulesToSaveWrapper`, `LNTuningLayer`) are fixed the same way, and `enable_adapters`'s signature gained an optional `inference_mode: bool = False` parameter (default preserves prior behavior for any external caller that doesn't pass it).

### Tests

Added one test parametrized over the existing `TEST_CASES` matrix (`test_disable_adapters_exiting_context_preserves_inference_mode`, covering every tuner type in that suite — LoRA, LoHa, LoKr, AdaLoRA, HiRA, BOFT, GLoRA, LN Tuning, UniLoRA, VeRA, WaveFT, VBLoRA, X-LoRA, TinyLoRA, etc.), plus two dedicated tests for `modules_to_save` and `trainable_token_indices` mirroring the existing `test_requires_grad_modules_to_save_disabling`/`test_requires_grad_follows_inference_mode_trainable_token_indices` tests but with `inference_mode=True`.

Per the contribution guide, implemented as a failing test first: all three new tests fail on `main` (321 failures total across the parametrized matrix) and pass after the fix.

Ran the following locally (CPU, Python 3.12):
- `pytest tests/test_custom_models.py` (full file): 15347 passed, 676 skipped, 4 xfailed (the 4 xfail are pre-existing, unrelated known limitations, e.g. `test_loading_model_requires_grad_set_correctly_switch_inference_mode`). One `TestDynamicDispatch::test_training_works` failure appeared under `-n auto`; it does not touch `enable_adapters`/`inference_mode` at all and passes consistently in isolation (both single-process and under xdist scoped to just that class), so it looks like a pre-existing xdist test-isolation flake (the suite already emits a `PytestRemovedIn10Warning: Class-scoped fixture defined as instance method is deprecated` for that class) rather than a regression from this change.
- `pytest tests/test_custom_models.py::TestRequiresGrad`: 175 passed, 2 skipped, 4 xfailed.
- `pytest tests/test_tuners_utils.py`: 219 passed, 16 skipped.
- `pytest tests/test_other.py tests/test_trainable_tokens.py`: 107 passed.
- `pytest tests/test_mixed.py`: 32 passed, 1 skipped.
- `make style` (`ruff check` + `ruff format --check`, pinned `ruff~=0.15.12`): clean.

## AI disclosure

I used AI assistance (Claude) to help trace the root cause, identify the two duplicated occurrences of the same bug, and draft the fix and tests. I reviewed every changed line, understand the fix and why it's needed, and ran all tests listed above myself.

## Coordination

Posted [this comment](https://github.com/huggingface/peft/issues/3447#issuecomment-5010032441) on the issue confirming the repro and stating intent before opening this PR, per `.ai/AGENTS.md`.

### CODE REVIEW

**Review by ErenAta16 — COMMENTED** [review]
Reviewed this against the root cause from #3447 (mine). This is the right fix, and it goes further than what I'd traced: I'd only caught `BaseTunerLayer.enable_adapters` itself, but this correctly finds the two duplicated copies in `ModulesToSaveWrapper`/`TrainableTokensWrapper` (`peft/utils/other.py`) that had the identical unconditional-unfreeze bug, plus threads `inference_mode` through `UniLoraLayer`/`LNTuningLayer` so their extra trainable state (`unilora_theta_d`, the LN-tuning params) doesn't get unfrozen either.

The `all(self.peft_config[name].inference_mode for name in module.active_adapters if name in self.peft_config)` choice in `_enable_adapter_layers` is the right call for the mixed-adapter case: if even one active adapter isn't in inference_mode, the module should stay trainable, so requiring *all* of them to be frozen before treating the module as frozen is the conservative, correct direction to default on ambiguity.

Test coverage matches the actual bug shape well. `test_requires_grad_modules_to_save_disabling_inference_mode` and the trainable-token-indices equivalent both go through the full disable/enable and context-manager round trip rather than just checking state right after `get_peft_model`, which is what actually would have caught this originally.

**Comment by BenjaminBossan** [comment]
Thanks for the PR but as discussed in [this comment](https://github.com/huggingface/peft/issues/3447#issuecomment-5010773818), it's a 'won't fix' for now.
